EIPS/eip-BN256_HashToCurve.md

```diff
@@ -0,0 +1,342 @@
+
+There is a similar proposal to add precompiled contracts for the
+elliptic curve BLS12-381 in
+[EIP-2537](./eip-2537.md);
+this proposal includes a deterministic map-to-curve algorithm.
+
+## Abstract
+There is currently no inexpensive way to perform BLS signature
+verification for arbitrary messages.
+This stems from the fact that there is no precompiled contract
+in the EVM for a hash-to-curve algorithm for the BN256 elliptic curve.
+The gas cost of calling a deterministic hash-to-curve algorithm
+written in Solidity is approximately that of one pairing check,
+although the latter requires an order of magnitude
+more computation.
+This EIP seeks to remedy this by implementing an algorithm
+for hashing to BN256 G1 curve, which would reduce the cost of
+signature verification to essentially that of the pairing check
+precompiled contract.
+We also include a hash-to-curve algorithm for the BN256 G2 group.
+
+## Motivation
+
+The precompiled contracts in
+[EIP-198](./eip-198.md) and
+[EIP-1108](./eip-1108.md)
+increased usage of cryptographic operations in the EVM
+by reducing the gas costs.
+In particular, the cost reduction from
+[EIP-1108](./eip-1108.md)
+helps increase the use of SNARKs in Ethereum
+via an elliptic curve pairing check;
+however, a hash-to-curve algorithm enabling arbitrary
+BLS signature verification on BN256 in the EVM was noticeably missing.
+There is interest in having a precompiled contract which would allow
+for signature verification, as noted
+[here](https://ethereum-magicians.org/t/pre-compile-for-bls/3973).
+If added, [EIP-2537](./eip-2537.md)
+would enable BLS signature verification on the BLS12-381 elliptic curve;
+however, no precompiled contracts currently exist
+for BLS12-381 elliptic curve operations.
+
+At this time, we are able to perform addition, scalar multiplication,
+and pairing checks in BN256.
+Reducing these costs in
+[EIP-1108](./eip-1108.md)
+made [ETHDKG](https://github.com/PhilippSchindler/ethdkg),
+a distributed key generation protocol in Ethereum,
+less expensive.
+ETHDKG by itself is useful; however, what it is lacking is
+the ability to verify arbitrary BLS signatures.
+Creating group signatures by aggregating partial signatures
+is one goal of a DKG protocol.
+The DKG enables the computation of partial signatures to be
+combined into a group signature offline, but there is no
+easy way to verify partial signatures or group signatures
+in the EVM.
+
+In order to perform BLS signature validation, a hash-to-curve
+function is required.
+While the MapToGroup method initially discussed in the original BLS
+[paper](https://crypto.stanford.edu/~dabo/pubs/abstracts/weilsigs.html)
+works in practice, the nondeterministic nature of the algorithm
+leaves something to be desired, as we would like to bound
+the overall computational cost in the EVM.
+A deterministic method for mapping to BN curves is given
+[here](https://www.di.ens.fr/~fouque/pub/latincrypt12.pdf);
```

Comment:
Same comment as above, can we include this directly as an asset?

Reply:
Same comment for all of the other papers linked.

Reply:
I have made local copies of this paper.

Reply:
MadNet paper has been added. This should take care of all references.